docs/Roles and Access Control/3.API Keys.md

```diff
@@ -33,3 +35,51 @@ To create and configure access control for API keys:
+
+### How It Works
+
+* Let's consider that AI DIAL Core can receive an initial request from AI DIAL Chat or any of its clients. Such request we will call **root** requests. Example: `AI DIAL Chat --> AI DIAL Core`
+* AI DIAL Core can perform a series of requests to realize the initial request - let's call them **non-root** requests. Example: `AI DIAL Chat --> AI DIAL Core --> Application A --> AI DIAL Core --> Application B...`
+
+**Root Request**:
+
+1. AI DIAL Core checks that the caller (chat user or application) has enough permissions and limits (based on the [configuration](#api-keys)) to perform the request and generates a new per-request key specifically for this request.
+2. AI DIAL Core stores an additional data associated with the per-request key either in memory or in Redis. The data may include but is not limited to: JWT of the user or API key of the application, `trace-id`, `core-parent-span-id`, list of the user files attached to the request.
+3. AI DIAL Core passes the per-request key to the application or model adapter in the API-key header. The application or model adapter then send the per-request key back to AI DIAL Core using the same API-key header for model requests or file storage API requests.
+4. AI DIAL Core invalidates the per-request key when the request is completed.
+
+**Non-root Requests**:
+
+1. For all non-root requests, AI DIAL Core checks, that the **incoming** per-request key is valid and known (the one from the root request).
+2. It checks the permissions of the caller and then generates and saves a new **outgoing** per-request key. This key is also associated with specific data: all data from the root request (its parent request) and the necessary data for the current request.
+3. AI DIAL Core passes this new outgoing per-request key in the request header. The application or model adapter then send the per-request key back to AI DIAL Core using the same API-key header for model requests or file storage API requests.
+4. AI DIAL Core invalidates the outgoing per-request key when the request is completed.
+
+### Calculating Costs & Limits
+
+Per-request keys can be used to attribute costs and limits incurred by a request to a model to the user or application that initiated the request.
+
+To attribute costs and limits to the request originator, `traceparent` should be included in the request header. `trace-id` is saved throughout the lifespan of the request in memory or in Redis.
+
+The application can as well propagate a `traceparent` header for the tracing feature. If the value of the `trace-id` propagated in the `traceparent` header does not match with the `trace-id` stored in the additional data associated with the per-request key, the application will receive 400 Bad request error.
+
+### Gathering Statistics
+
+For calculating statistics, AI DIAL Core uses `trace-id` and `core-parent-span-id` that are stored in the additional data linked to the per-request key.
+
+### Files Sharing
```

Comment:
Please add sections for auto sharing input and output files.
There is a vital info missed:
```
To provide the path of the folder for output files, we suggest to extent the [GET /v1/bucket](https://epam-rail.com/dial_api#/paths/~1v1~1bucket/get) response with an additional field - appdata field.
If the [GET /v1/bucket](https://epam-rail.com/dial_api#/paths/~1v1~1bucket/get) request is done with a Per-request key the response will contain appdata field with the path to the shared application folder, including the User's bucket and the correct deployment-id for the application as registered in the Core configs.

GET /v1/bucket
{
  "bucket": "{application-bucket-id}",
  "appdata": "{user-bucket-id}/appdata/{deployment-id}",
}
```

Reply:
added